core/src/main/java/org/apache/iceberg/avro/InternalReader.java

```diff
@@ -0,0 +1,251 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.avro;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Supplier;
+import org.apache.avro.LogicalType;
+import org.apache.avro.LogicalTypes;
+import org.apache.avro.Schema;
+import org.apache.avro.io.DatumReader;
+import org.apache.avro.io.Decoder;
+import org.apache.iceberg.StructLike;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.types.Type;
+import org.apache.iceberg.types.Types;
+import org.apache.iceberg.util.Pair;
+
+/**
+ * A reader that produces Iceberg's internal in-memory object model.
+ *
+ * <p>Iceberg's internal in-memory object model produces the types defined in {@link
+ * Type.TypeID#javaClass()}.
+ *
+ * @param <T> Java type returned by the reader
+ */
+public class InternalReader<T> implements DatumReader<T>, SupportsRowPosition {
+
+  private final Types.StructType expectedType;
+  private final Map<Integer, Class<? extends StructLike>> typeMap = Maps.newHashMap();
+  private final Map<Integer, Object> idToConstant = ImmutableMap.of();
```

Comment:
Do we actually need this field if it is immutable? Will it change once we add default values?

Reply:
Yes, this is going to change when we have default values and will be populated in the next PR.

Reply:
Do you plan another PR to add V3 default-values support (just curious) ?

Reply:
@jbonofre, sorry I missed this earlier.

I was wrong in my last comment. Default values come directly from the expected schema and support for them was added when I rebased this on top of @wmoustafa's recent PR (#9502). This constant map is here as a placeholder for when we add the ability to pass in constants that override fields, which is how we implement partition values. Those must be stored in data files, but are not read.